library/core/src/ops/deref.rs

```diff
@@ -3,15 +3,44 @@
 /// In addition to being used for explicit dereferencing operations with the
 /// (unary) `*` operator in immutable contexts, `Deref` is also used implicitly
 /// by the compiler in many circumstances. This mechanism is called
-/// ['`Deref` coercion'][more]. In mutable contexts, [`DerefMut`] is used.
-///
-/// Implementing `Deref` for smart pointers makes accessing the data behind them
-/// convenient, which is why they implement `Deref`. On the other hand, the
-/// rules regarding `Deref` and [`DerefMut`] were designed specifically to
-/// accommodate smart pointers. Because of this, **`Deref` should only be
-/// implemented for smart pointers** to avoid confusion.
-///
-/// For similar reasons, **this trait should never fail**. Failure during
+/// ['`Deref` coercion'][more]. In mutable contexts, [`DerefMut`] is used and
+/// mutable deref coercion similarly occurs.
+///
+/// Deref coercion is a powerful language feature and has far-reaching
+/// implications for every type implementing `Deref`: the compiler will often
+/// implicitly call `Deref::deref`, as described [below][more]. For this reason,
+/// `Deref` should not be implemented unless all of the implications of coercion
+/// are understood and desired. Implementing `Deref` is most commonly desirable
+/// when a type
+///
+/// 1. should transparently behave like a contained value; and
+/// 2. does not have a rich set of methods in of itself, to avoid conflict or
+///    confusion with methods on the `Target` type.
+///
+/// For example, [`Box<T>`][box] is perhaps the canonical example for `Deref`:
+/// a `Box<T>` value **is** a `T` value that is heap-allocated, so deref
+/// coercion intuitively makes `Box<T>` act like `T`. `Box<T>`
+/// also has very few methods (though several associated functions), so it is
+/// unlikely that users will find conflict with methods on their target type.
+/// [`RefCell<T>`][refcell] also contains a value of `T`, but while it could
+/// implement `DerefMut`, it does not since there is a rich API of methods to
```

Comment:
Could it? We need to return a lock (refcell) guard, and Deref's current API does not allow returning an owned value.

I think the reason we're not doing it isn't because of the rich API of methods though, rather specifically because when you take (and release) locks is often important, and hiding that behind an implicit coercion makes it much harder to audit such code, including in code reviews.

Reply:
I chose `DerefMut` here rather than `Deref` specifically. `RefCell::get_mut` has exactly the same signature as `DerefMut::deref_mut` (modulo generics/associated types) and requires no runtime checks or locking because it takes `&mut self`. I posit that a sufficient reason for not implementing `DerefMut` here is the rich-API-problem (though there may be other sufficient conditions, like the value in more explicit APIs because locking _might_ be involved with `RefCell`).

This was the best example I came up with in the standard library (though I could easily have missed a better one).